### DISCUSSION TITLE
FrustumIntersection.compute

### DISCUSSION THREAD

**Discussion by blueRaining** [heremaps/harp.gl#2267] [Ideas]
for (const tileEntry of this.m_rootTileKeys) {
            const tileKey = tileEntry.tileKey;
            const offset = tileEntry.offset;
            // We even check the root tiles against the frustum b/c it can happen that
            // computeRequiredInitialRootTileKeys is producing false positives.
            const tileKeyEntry = this.getTileKeyEntry(tileKey, offset, tilingScheme, cache, minGeometryHeight, maxGeometryHeight, useElevationRangeSource ? elevationRangeSource : undefined);
            if (tileKeyEntry !== undefined) {
              for (const zoomLevel of uniqueZoomLevels) {
                if (tileKeyEntry.tileKey.level !== zoomLevel) {
                  continue;
              }
                    const tileKeyEntries = this.m_tileKeyEntries.get(zoomLevel);
                    tileKeyEntries.set(Utils_1.TileOffsetUtils.getKeyForTileKeyAndOffset(tileKey, offset), tileKeyEntry);
                }
            }
        }
        const workList = [...this.m_rootTileKeys.values()];
        while (workList.length > 0) {
            const tileEntry = workList.pop();
            if (tileEntry === undefined) {
                break;
            }
            // Stop subdivision if hightest visible level is reached
            const tileKey = tileEntry.tileKey;
            const offset = tileEntry.offset;
            const subdivide = dataSources.some((ds, i) => ds.shouldSubdivide(zoomLevels[i], tileKey));
            if (!subdivide) {
                continue;
            }
            // Stop subdivision if area of tile is too small(mixed LOD only)
            if (this.m_enableMixedLod && tileEntry.area < targetTileArea) {
                continue;
            }
            const tileKeyAndOffset = Utils_1.TileOffsetUtils.getKeyForTileKeyAndOffset(tileKey, offset);
            // delete parent tile key from applicable zoom levels
            // for (const zoomLevel of uniqueZoomLevels) {
            //     if (tileKey.level >= zoomLevel) {
            //         continue;
            //     }
            // 
            //     const tileKeyEntries = this.m_tileKeyEntries.get(zoomLevel);
            //     tileKeyEntries.delete(tileKeyAndOffset);
            // }
            for (const subTileKey of tilingScheme.getSubTileKeys(tileKey)) {
                const subTileEntry = this.getTileKeyEntry(subTileKey, offset, tilingScheme, cache, minGeometryHeight, maxGeometryHeight, useElevationRangeSource ? elevationRangeSource : undefined);
                if (subTileEntry !== undefined) {
                    for (const zoomLevel of uniqueZoomLevels) {
                        if (subTileEntry.tileKey.level !== zoomLevel) {
                            continue;
                        }
                        const subTileKeyAndOffset = Utils_1.TileOffsetUtils.getKeyForTileKeyAndOffset(subTileKey, offset);
                        this.m_tileKeyEntries
                            .get(zoomLevel)
                            .set(subTileKeyAndOffset, subTileEntry);
                    }
                    workList.push(subTileEntry);
                }
            }
        }

can we use  'subTileEntry.tileKey.level !== zoomLevel' and   'tileKeyEntry.tileKey.level !== zoomLevel' to reduce for loop in workList

**Comment by nzjony** [comment]
@blueRaining , where did you get this code snippet?

Please make a PR so I can better understand what you mean... even if the PR is just some comments. Otherwise I need to search for the code you added above into our code and do a manual mental diff and that is too much effort. I don't mean to be rude, I just want to help you better and make best use of my time.

**Reply by blueRaining** [reply]
![image](https://user-images.githubusercontent.com/41565968/128275234-06c36205-109d-464a-b0c4-859c3935d987.png)
![image](https://user-images.githubusercontent.com/41565968/128275253-00499365-062b-49c7-a568-c9f7a99b1b60.png)
![image](https://user-images.githubusercontent.com/41565968/128275270-9087f7e1-7ef1-4d4f-80e4-6bda33328657.png)
those are my changes to reduce for loop

**Comment by nzjony** [comment]
hey @blueRaining , I'm a bit busy now, when I have some time, I come back and take a look.

It has been a while since I last looked at this code, so I need to refresh what is going on... unfortunately it is not so easy to understand.

Can you run the tests locally: `yarn test`, do they pass?

**Reply by blueRaining** [reply]
Yes it is correct

**Reply by nzjony** [reply]
@blueRaining , can you please make a PR so I can test locally?
